### ISSUE TITLE
Allow users to click through the window

### ISSUE THREAD

**Issue by Ayplow** [rust-windowing/winit#1434]
For overlays that draw over other windows, it's important that the user is able to interact with the applications "below" the overlay. Some programs that make use of this are [f.lux](https://justgetflux.com/) and [Discord's Overlay](https://discordapp.com).

Somewhere to start with an implementation:
- [Windows' `WS_EX_TRANSPARENT` window style](https://www.codeproject.com/Articles/12365/How-to-make-a-transparent-and-click-through-window)
- [X11's Nonrectangular Window Shape Extension Protocol](https://www.x.org/releases/X11R7.7/doc/xextproto/shape.html)
- [NSWindow's `ignoresMouseEvents` flag](https://developer.apple.com/documentation/appkit/nswindow/1419354-ignoresmouseevents)

**Comment by samhedin** [comment]
I would love this feature. On windows it, as you say, just seems to be a matter of setting `WS_EX_TRANSPARENT`. It's already used in some form here, with the use of `winuser::WS_EX_TRANSPARENT` https://docs.rs/winit/0.22.1/x86_64-pc-windows-msvc/src/winit/platform_impl/windows/event_loop.rs.html
I have no idea how to start working on this, but I would be interested in putting in some effort if someone would point me in the right direction.

**Comment by samhedin** [comment]
I tried my hand at this and from what it looks to me, it shouldn't be worse than setting the bit (for windows) in ´windows/window_state.rs´. As a test, I added
`        style_ex |= WS_EX_COMPOSITED;
        style_ex |= WS_EX_TRANSPARENT;
`
to the bottom of the function `to_window_styles` as well as commenting out `| winuser::CS_OWNDC` in `window.rs` as per the note in https://docs.microsoft.com/en-us/windows/win32/winmsg/extended-window-styles

This did not work, neither did any combination of these settings work. I might be flailing around a bit randomly - but most resources I see (including the one linked in op) seem to say that this should be all it takes. I might be missing something basic here.

**Comment by kchibisov** [comment]
not sure if it'll work on Wayland, but there's no standard protocol for overlays either now on it. Though, I wonder if winit should support overlays, since they usually use special set of protocols on each platform IIRC.

**Comment by kazzix14** [comment]
I also would love this future. think [this](https://www.x.org/releases/X11R7.7/doc/compositeproto/compositeproto.txt) can be used?
> 3.2 Composite Overlay Window
Version 0.3 of the protocol adds the Composite Overlay Window,

**Comment by maroider** [comment]
I actually experimented with this almost _exactly_ a year ago, today. I got it _sort of_ working on Windows, but it doesn't work on full-screen games/applications. The repository is [here](https://github.com/maroider/overlay), and the interesting bits are [in this file](https://github.com/maroider/overlay/blob/1a80a30d6ef8e6b2fe6fa273a2cc2b472a3b2e51/src/os.rs).

It's important here to acknowledge that there are two types of overlays: Internal overlays, which hook D3D/OpenGl/Vulkan functions and render themselves there, and external overlays, which create a window which puts itself on top of another window.
Steam and Discord's overlays are likely internal, given that they both work with games in exclusive full-screen, and that they both inject one or more DLLs into the game being overlaid.

I'm also concerned about how the various input APIs would interact with these "click-trough" overlay windows. I have no idea how you'd robustly block the window beneath the overlay from using a gamepad while the overlay wants to use it exclusively. I've already noticed that at least one game I play ignores window focus WRT gamepad input, so they still allow me to control the game while I'm using my browser. An internal overlay could simply hook the relevant functions.

Exposing platform-specific APIs which let you set the correct flags and such to create an external overlay is probably feasible, but support for creating external overlays in Winit proper is not something I'm convinced is advisable.

On a related note: I'm actually working on a prototype for a Winit-like crate for internal overlays. I'll probably have something vaguely presentable, and probably Windows-only, at the end of January.
